test/ai/TestGenerateRandomMove.java

```diff
@@ -0,0 +1,349 @@
+		this.catLoverAi = new Ai(Owner.Player2, new Game(new BoardState(catLoverPieces)));
+	}
+
+	public void randomStressTest(HashMap<Object, Double> expectedPercentages, Generater method) {
+		HashMap<Object, Double> countMap = new HashMap<Object, Double>();
+		for (int i = 0; i < ITERATION_SIZE; i++) {
+			Object returnValue = method.generate();
+			if (!countMap.containsKey(returnValue)) {
+				countMap.put(returnValue, 0.0);
+			}
+			countMap.put(returnValue, countMap.get(returnValue) + 1);
+		}
+
+		// convert count to percentages
+		for (Object obj : countMap.keySet()) {
+			countMap.put(obj, round(countMap.get(obj) / ITERATION_SIZE, 2));
+		}
+
+		for (Object obj : countMap.keySet()) {
+			double expectedPercent = expectedPercentages.get(obj);
+			double expectedLow = Math.max(expectedPercent - RANDOM_MARGIN, Math.min(Double.MIN_NORMAL, expectedPercent));
+			double expectedHigh = Math.min(expectedPercent + RANDOM_MARGIN, Math.max(100.0 - Double.MIN_NORMAL, expectedPercent));
+			double percent = countMap.get(obj);
+			String errorString = obj.toString() + " was outside of the expected range of " + (expectedPercent * 100)
+					+ "% +- " + (RANDOM_MARGIN * 100) + "% with a percentage of " + (percent * 100) + "%";
+			if (expectedLow > percent || percent > expectedHigh) {
+				StackTraceElement methodName = Thread.currentThread().getStackTrace()[2];
+				System.err.println("FAIL " + methodName);
+				// System.err.println(countMap);
+				System.out.println();System.err.println();
+			}
+			assertTrue(errorString, expectedLow <= percent && percent <= expectedHigh);
+		}
+	}
+
+	public void randomStressTest(double expectedPercentages, Generater method) {
+		HashMap<Object, Double> countMap = new HashMap<Object, Double>();
+		for (int i = 0; i < ITERATION_SIZE; i++) {
+			Object returnValue = method.generate();
+			if (!countMap.containsKey(returnValue)) {
+				countMap.put(returnValue, 0.0);
+			}
+			countMap.put(returnValue, countMap.get(returnValue) + 1);
+		}
+
+		// convert count to percentages
+		for (Object obj : countMap.keySet()) {
+			countMap.put(obj, round(countMap.get(obj) / ITERATION_SIZE, 2));
+		}
+
+		for (Object obj : countMap.keySet()) {
+			double expectedPercent = expectedPercentages;
+			double expectedLow = Math.max(expectedPercent - RANDOM_MARGIN, Math.min(Double.MIN_NORMAL, expectedPercent));
+			double expectedHigh = Math.min(expectedPercent + RANDOM_MARGIN, Math.max(100.0 - Double.MIN_NORMAL, expectedPercent));
+			double percent = countMap.get(obj);
+			String errorString = obj.toString() + " was outside of the expected range of " + (expectedPercent * 100)
+					+ "% +- " + (RANDOM_MARGIN * 100) + "% with a percentage of " + (percent * 100) + "%";
+			if (expectedLow > percent || percent > expectedHigh) {
+				StackTraceElement methodName = Thread.currentThread().getStackTrace()[2];
+				System.err.println("FAIL " + methodName);
+				// System.err.println(countMap);
```

Comment:
to print out, when push and pulls, equals and hashcode functions are implemented